Documentation/guides/messages/apt2264.md

```diff
@@ -0,0 +1,53 @@
+---
+title: Xamarin.Android error APT2264
+description: APT2264 error code
+ms.date: 12/16/2022
+---
+# Xamarin.Android error APT2264
+
+## Issue
+
+The tool `aapt2` is unable to resolve one of the files it was passed.
+This is generally caused by the path being longer than the Maximum Path
+length allowed on windows.
+
+## Solution
```

Comment:
Can we suggest enabling long paths as the 1st solution?

https://learn.microsoft.com/windows/win32/fileio/maximum-file-path-limitation?tabs=registry#enable-long-paths-in-windows-10-version-1607-and-later

Reply:
As per the docs, *two* things must be done: the registry must be updated to set `HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\FileSystem\LongPathsEnabled` -- and I wonder if that could/would break anything -- and ***secondly***:

> * The [application manifest](https://learn.microsoft.com/en-us/windows/win32/sbscs/application-manifests) must also include the `longPathAware` element.
> ```xml
> <application xmlns="urn:schemas-microsoft-com:asm.v3">
>     <windowsSettings xmlns:ws2="http://schemas.microsoft.com/SMI/2016/WindowsSettings">
>         <ws2:longPathAware>true</ws2:longPathAware>
>     </windowsSettings>
> </application>
> ```

We do not currently ship an `aapt2.exe.config` manifest file.

Additionally, it may be moot; the next section in docs states:

> These are the directory management functions that no longer have MAX_PATH restrictions if you opt-in to long path behavior: CreateDirectoryW, CreateDirectoryExW GetCurrentDirectoryW RemoveDirectoryW SetCurrentDirectoryW.
>
> These are the file management functions that no longer have MAX_PATH restrictions if you opt-in to long path behavior: CopyFileW, CopyFile2, CopyFileExW, CreateFileW, CreateFile2, CreateHardLinkW, CreateSymbolicLinkW, DeleteFileW, FindFirstFileW, FindFirstFileExW, FindNextFileW, GetFileAttributesW, GetFileAttributesExW, SetFileAttributesW, GetFullPathNameW, GetLongPathNameW, MoveFileW, MoveFileExW, MoveFileWithProgressW, ReplaceFileW, SearchPathW, FindFirstFileNameW, FindNextFileNameW, FindFirstStreamW, FindNextStreamW, GetCompressedFileSizeW, GetFinalPathNameByHandleW.

Are these used?

```
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Xamarin\Android> dumpbin /imports  aapt2.exe  | findstr File
File Type: EXECUTABLE IMAGE
                   8C CreateFileMappingA
                   90 CreateFileMappingW
                  1E9 GetFileAttributesA
                  1EE GetFileAttributesW
                  1F0 GetFileInformationByHandle
                  1F5 GetFileSizeEx
                  1F7 GetFileType
                  216 GetModuleFileNameA
                  27E GetSystemTimeAsFileTime
                  362 MapViewOfFile
                  3CF ReadFile
                  43B SetEndOfFile
                  44D SetFilePointer
                  4BD UnmapViewOfFile
```

Odd is `GetFileAttributesA`, which presumably has `MAX_PATH` limits, but `GetFileAttributesW` is *also* used, so when is `GetFileAttributesA` used vs. `GetFileAttributesW`? ¯\\_(ツ)_/¯.

`GetModuleFileNameA` is also concerning, so it would not surprise me if `aapt2.exe` didn't work with long filenames even if we set the registry entry *and* provided an appropriate `aapt2.exe.config` manifest.

This requires testing, but I suspect that there's no point in suggesting the enabling of long paths, as `aapt2` won't work with them.

Reply:
`aapt2` [uses](https://android.googlesource.com/platform/frameworks/base/+/refs/tags/android-13.0.0_r24/tools/aapt2/util/Files.cpp#52) this [function](https://android.googlesource.com/platform/system/libbase/+/refs/tags/android-13.0.0_r24/utf8.cpp#162) to convert UTF-8 file names to Windows Unicode

The android base library also [has wrappers](https://android.googlesource.com/platform/system/libbase/+/refs/tags/android-13.0.0_r24/utf8.cpp#181) around the standard C I/O functions which add support for UTF-8 strings.

Reply:
This strongly implies that if we provide an appropriate `aapt2.exe.config`, we can get long file name support *if*/when the registry setting is enabled.

Reply:
Maybe we should just check if you need it? My memory in the past, I just changed the registry key and everything worked.

But I don't remember if I was hitting an issue with `aapt2` or a .NET process (which would probably opt into long paths).

Reply:
@dellis1972: if this PR were "green" and otherwise ready to merge, I think `aapt2.exe.config` could be done as a separate PR.  Given that this PR *isn't* green and ready to merge, might as well test locally and see if that works?

Reply:
I think you might have to test an "almost long path", where only aapt2 would get long path errors?

Reply:
So I did some more testing and it seems that `aapt2` was able to switch to log paths with only the registry setting change (not aapt2.exe.config file). 

I know this because the diagnostic output was putting the `\\?\` stuff at the case of the filenames. So I don't think we need the config file. I will update the docs to mention the registry change though.